File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        vector<int> dp(prices.size(), 0);
        for(int i=1; i<prices.size(); i++) {
            dp[i] = dp[i-1];
            for(int j=i-1; j>=0; j--) {
                if(prices[j] < prices[i]) {
                    int max_profit = max(dp[i-1], dp[j] + prices[i]-prices[j]);
                    dp[i] = max(dp[i], max_profit);
                }
            }
        }
        // for(int n: dp) {
        //     cout << n << " ";
        // }
        return dp[prices.size()-1];
    }
};

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[][] dp = new int[n+1][2];

        for(int i=n-1; i>=0; i--) {
            for(int canBuy=1; canBuy>=0; canBuy--) {
                if(canBuy == 1) {
                    int buy = -prices[i] + dp[i+1][0];
                    int skip = dp[i+1][1];
                    dp[i][1] = Math.max(buy, skip); 
                }
                else {
                    int sell = prices[i] + dp[i+1][1];
                    int skip = dp[i+1][0];
                    dp[i][0] = Math.max(sell, skip);
                }
            }
        }

        return dp[0][1];
    }
}